staging/src/k8s.io/apiserver/pkg/endpoints/handlers/apply.go

```diff
@@ -17,164 +17,80 @@ limitations under the License.
 
 import (
+	"errors"
 	"fmt"
 
-	"github.com/ghodss/yaml"
-	"k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/apimachinery/pkg/api/meta"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-	"k8s.io/apimachinery/pkg/apply"
-	"k8s.io/apimachinery/pkg/apply/parse"
-	"k8s.io/apimachinery/pkg/apply/strategy"
 	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/util/json"
-	"k8s.io/kube-openapi/pkg/util/proto"
+	"k8s.io/apiserver/pkg/endpoints/handlers/apply"
+	"sigs.k8s.io/structured-merge-diff/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/merge"
+	"sigs.k8s.io/structured-merge-diff/typed"
 )
 
 type applyPatcher struct {
 	*patcher
 
-	model proto.Schema
+	converter apply.TypeConverter
+	updater   merge.Updater
 }
 
-// TODO(apelisse): workflowId needs to be passed as a query
-// param/header, and a better defaulting needs to be defined too.
-const workflowId = "default"
+const applyManager = "apply"
 
-func (p *applyPatcher) convertCurrentVersion(obj runtime.Object) (map[string]interface{}, error) {
-	vo, err := p.unsafeConvertor.ConvertToVersion(obj, p.hubGroupVersion)
+func (p *applyPatcher) applyPatchToCurrentObject(newObject, currentObject runtime.Object) (runtime.Object, error) {
+	vo, err := p.unsafeConvertor.ConvertToVersion(currentObject, p.hubGroupVersion)
 	if err != nil {
-		return nil, err
-	}
-	return runtime.DefaultUnstructuredConverter.ToUnstructured(vo)
-}
-
-func (p *applyPatcher) extractLastIntent(obj runtime.Object, workflow string) (map[string]interface{}, error) {
-	accessor, err := meta.Accessor(obj)
-	if err != nil {
-		return nil, fmt.Errorf("couldn't get accessor: %v", err)
-	}
-	last := make(map[string]interface{})
-	// TODO: use the managedFields correctly
-	if _, ok := accessor.GetManagedFields()[workflow]; ok {
-		if err := json.Unmarshal([]byte(accessor.GetManagedFields()[workflow].APIVersion), &last); err != nil {
-			return nil, fmt.Errorf("couldn't unmarshal managedFields field: %v", err)
-		}
+		return nil, fmt.Errorf("failed to convert object to version: %v", err)
 	}
-	return last, nil
-}
-
-func (p *applyPatcher) getNewIntent() (map[string]interface{}, error) {
-	patch := make(map[string]interface{})
-	if err := yaml.Unmarshal(p.patchBytes, &patch); err != nil {
-		return nil, fmt.Errorf("couldn't unmarshal patch object: %v (patch: %v)", err, string(p.patchBytes))
-	}
-	return patch, nil
-}
-
-func (p *applyPatcher) convertResultToUnversioned(result apply.Result) (runtime.Object, error) {
-	voutput, err := p.creater.New(p.kind)
+	current, err := runtime.DefaultUnstructuredConverter.ToUnstructured(vo)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create empty output object: %v", err)
+		return nil, fmt.Errorf("failed to convert to unstructured: %v", err)
 	}
 
-	err = runtime.DefaultUnstructuredConverter.FromUnstructured(result.MergedResult.(map[string]interface{}), voutput)
+	managed, err := apply.DecodeObjectManagedFields(current)
```

Comment:
Yeah, this doesn't have to work on an unstructured, you can use the Accessor on any type of runtime.Object